### ISSUE TITLE
Typescript: Context message does not contain text

### ISSUE THREAD

**Issue by german-bortoli** [telegraf/telegraf#1319]
Hello everyone, before starting this thread I would love to introduce my setup

* Telegraf.js Version:  4.0.1
* Node.js Version: 14
* Operating System: OSX or Linux

The main problem is when I am using TypeScript It's not possible to access on Text property from the message itself.

```typescript
  import { Context } from 'telegraf';

   // *** [More code...] ***

  bot.on(['text', 'edited_message'], (ctx: Context) => {
    const msg = ctx.message ?? ctx.editedMessage;
    console.log(msg.text) // We will read the message text into the console
  });
```

This will throw a typescript error described below.

`TS2339: Property 'text' does not exist on type '(New & NonChannel & TextMessage) | (New & NonChannel & DocumentMessage) | (New & NonChannel & AudioMessage) | ... 54 more ... | (Edited & ... 1 more ... & VoiceMessage)'.   Property 'text' does not exist on type 'New & NonChannel & DocumentMessage'.`

And this is my tsconfig.json

```json
{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Node 14",
  "compilerOptions": {
    "module": "commonjs",
    "lib": ["es2020"],
    "esModuleInterop": true,
    "removeComments": true,
    "allowSyntheticDefaultImports": true,
    "target": "es6",
    "noImplicitAny": true,
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "build",
    "baseUrl": ".",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "paths": {
      "*": [
        "node_modules/*",
        "src/types/*"
      ]
    }
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "__tests__"
  ]
}

```

**Comment by wojpawlik** [comment]
I don't know how to explain this simply... so I'll explain succinctly:

`edited_message` updates are a superset of `edited_message.text` (#1218) updates.

For now, use [`bot.guard`](https://telegraf.js.org/classes/telegraf.html#guard).

**Comment by dilastra** [comment]
hello))
This code solved your problem: 
```javascript
bot.on("text", (ctx: Context) => {
  const { text } = ctx.message as Message.TextMessage;
  if (text) {
    //do something
  }
});

bot.on("edited_message", (ctx: Context) => {
  //do something
});
```

The problem is that text message after edited may not be text message
For this reason, it is better to split it into two handlers

**Comment by wojpawlik** [comment]
You shouldn't write the type annotation for `ctx` there. It's correctly inferred as `GuardedContext<Context, { message: Message.TextMessage }>`, making the cast unnecessary.

**Comment by dilastra** [comment]
Type assertions is not write the type annotation.
https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

**Comment by HouCoder** [comment]
I'm having the same problem with the `document` property.

```
Property 'document' does not exist on type 'New & NonChannel & TextMessage & object'.ts(2339)
```

**Comment by EugeneDraitsev** [comment]
@wojpawlik if not every edited message is edited text message, does that mean  that the `text` property should exist at the type, but should be nullable or am I wrong?

**Comment by wojpawlik** [comment]
3.38 did that, and it resulted in non-null assertions everywhere.

**Comment by EugeneDraitsev** [comment]
So, how I can get `text`, `reply_to_message` and `entities` property of message in 4.0+?  
F.e. I have this logic:

```tsx
bot.use(async (ctx: Context, next) => {
  const { chat, message } = ctx
  const { message_id, reply_to_message } = message
  const text = message?.text
   if (isBotCommand(message?.entities)) {
      // some logic
   }
})
```
With 4.0+ error looks like 'You will never have `text`, `reply_to_message` and `entities` properties in message", but message obviously can have these fields. How can I workaround it and why I should do it ?

**Comment by wojpawlik** [comment]
Preferably, use [`.on('text')`](https://telegraf.js.org/classes/telegraf.html#on). If there's no `updateType` you want, use [`.guard`](https://telegraf.js.org/classes/telegraf.html#guard), or narrow down manually:

```ts
// 4.0
if (ctx.message === undefined) return next()
if(!('text' in ctx.message)) return next()
```

```ts
// 4.1 (hopefully)
if(!is.text(ctx.message)) return next()
```

Then TypeScript knows that `ctx.message.text` is present, and thus for example `ctx.message.photo` must be absent.

```ts
// 3.38 would require non-null assertions
bot.on('text', (ctx) => {
  console.log(ctx.message!.text!.toLowerCase())
})
```

**Comment by EugeneDraitsev** [comment]
I still can't get what is the problem to always have a `text` in message? You can use 3.38 version without non-null assertions:

```tsx
bot.use((ctx) => {
  console.log(ctx.message?.text) // or ctx.message?.text ?? '' if you need to pass a string
})
```

It looks like you're trying to remove possible fields form type, but these fields are exist it this type, based on [telegram-typings](https://github.com/sergeysova/telegram-typings/blob/master/javascript/index.d.ts) and on telegram official docs.

Also based on typescript handbook there is no any sense to remove optional properties from interfaces. There is makes sense to clarify types like this:

```tsx
bot.use((ctx: TextContext) => {
  console.log(ctx.message.text) // now you don't need to use optional chaining
})

// or

bot.on('text', (ctx) => {  // use TextContext by default for bot.text
  console.log(ctx.message.text)
})
```

My point here is that maybe you should consider not removing optional fields from interfaces and fix nullability with more specific contexts interfaces.

**Comment by wojpawlik** [comment]
Updated my 3.38 example. `bot.use((ctx: TextContext)` is a horrible idea.

**Comment by EugeneDraitsev** [comment]
Could you explain why?

```tsx
bot.use((ctx) => {
  console.log(ctx.message?.text?.toLowerCase()) // will work perfectly 
})
```

**Comment by wojpawlik** [comment]
Yes, it works as expected in 3.38, at a price:

```ts
// 3.38
bot.on('text', (ctx) => {
  console.log(ctx.message!.text!.toLowerCase())
})
```

`bot.use((ctx: TextContext)` would be a lie: pretending that `ctx.message.text` is available without verifying it.

**Comment by EugeneDraitsev** [comment]
Ok, I agree that `bot.use((ctx: TextContext)` looks like 'hack' and we shouldn't use it. But why don't  you like the optional chaining solution for `bot.use`? 

```tsx
bot.use((ctx) => { 
  console.log(ctx.message?.text?.toLowerCase()) // will work perfectly, but text should be an optional string for all messages
})
```

Users will be able to choose between `if (ctx.message === undefined) return next()` or `if(!is.text(ctx.message)) return next()` or just use `.? `

**Comment by wojpawlik** [comment]
I don't dislike it. I'm just unable to allow it without requiring non-null assertions everywhere else. PR welcome.

**Comment by EugeneDraitsev** [comment]
Ok, thanks for the explanations. I'll try to check it and maybe will create an PR

**Comment by wojpawlik** [comment]
Would this satisfy you?

```ts
// 4.1 hopefully
bot.use(ctx => {
  console.log(deunionize(ctx.message)?.text.toLowerCase())
})
```

I have 2 implementations already, sadly neither works on `Chat`, I can work on that.

**Comment by EugeneDraitsev** [comment]
It's good that we can have it, but it still feels a bit 'hacky'. I'll spend some time and check typings in repo, but if it's impossible to avoid non-null assertions, then your solution looks good

**Comment by wojpawlik** [comment]
Not perfect, but worked well on everything I threw at it:

```ts
type PropOr<T, P extends string | symbol | number, D> = T extends Partial<Record<P, infer V>> ? V : D

type UnionKeys<T> = T extends unknown ? keyof T : never

type Deunionize<T> = ([undefined] extends [T] ? undefined : never) | {
  [K in UnionKeys<T>]: PropOr<NonNullable<T>, K, undefined>
}

function deunionize<T extends object | undefined>(t: T) {
  return t as Deunionize<T>
}
```

I'll ship it in 4.1 so that people can improve it via pull requests.

**Comment by wojpawlik** [comment]
https://github.com/tdlib/telegram-bot-api/issues/93 would allow

```js
if (ctx.message?.type !== 'text') return next()
```

**Comment by YakovL** [comment]
in 4.6.0, `ctx.message?.type` gives

> Property 'type' does not exist on type 'New & NonChannel & Message'.
  Property 'type' does not exist on type 'New & NonChannel & ChannelChatCreatedMessage'.ts(2339)

I'm looking for a consistent method to decompose a message (text, photos, etc), but the only way I found working so far is

    const {
        text,
        forward_from_chat, forward_from_message_id,
        photo, caption, caption_entities,
        // still have to find out what props will be here for videos, documents etc
    } = ctx.message as any

and IDE doesn't give any helpful suggestions for `ctx.message.` so to figure what will be available for a video, I have to send a video and `console.log(ctx.message)`, same for audio, document, link etc.

Should I create a new issue? Look like this is tightly related.

**Comment by MKRhere** [comment]
@YakovL See https://github.com/telegraf/telegraf/issues/1388#issuecomment-791573609

**Comment by igriu** [comment]
Maybe this will be helpful to someone.
After trying several workarounds to get the text from the ctx message, I have found a possible solution.

```
import { Context, Telegraf, Telegram } from 'telegraf';
import { Update } from 'typegram'

bot.action('channel_click',  async (ctx) => {    
   let upd = ctx.update ;
   try{
    if('message' in upd.callback_query){
        if("text" in upd.callback_query.message){
        console.log(upd.callback_query.message?.text)
        }   
    }
   }
   catch(error){
     console.log(error)
   }
})
```
My bot posts messages in a Channel, the messages have a keyboard with one button attached.
The above action method reacts to the button click (callback_data: 'channel_click').

In my case (context), the message text is in **ctx.update.callback_query.message.text**.
The ctx is of type `Context<Update>`.

Also to avoid type errors in code, I have switched to **"strict": false** type checking in tsconfig.json
Does not look right but it is working.
